src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -545,14 +625,41 @@ private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
             return new ValueTask<int>(vts, vts.Version);
         }
 
-        private static NativeOverlapped GetNativeOverlapped(long fileOffset)
+        private static unsafe NativeOverlapped* GetNativeOverlappedForSynchronousOperation(SafeFileHandle handle, long fileOffset, NativeOverlapped stackAllocated = default)
         {
-            NativeOverlapped nativeOverlapped = default;
+            NativeOverlapped* result;
+            if (handle.IsAsync)
+            {
+                handle.EnsureThreadPoolBindingInitialized();
+
+                // After SafeFileHandle is bound to ThreadPool, we need to use ThreadPoolBinding
+                // to allocate a native overlapped and provide a valid callback.
+                // Since we really don't care about the callback (because this is sync IO for async handle)
+                // and we are going to wait on WaitHandle anyway, we pass ThreadPoolBinding as a state.
+                // The callback is going to use it to free the native overlapped.
+                result = handle.ThreadPoolBinding!.AllocateNativeOverlapped(s_callback, handle.ThreadPoolBinding, null);
+            }
+            else
+            {
+                result = &stackAllocated;
+            }
+
             // For pipes the offsets are ignored by the OS
-            nativeOverlapped.OffsetLow = unchecked((int)fileOffset);
-            nativeOverlapped.OffsetHigh = (int)(fileOffset >> 32);
+            result->OffsetLow = unchecked((int)fileOffset);
+            result->OffsetHigh = (int)(fileOffset >> 32);
+
+            return result;
+        }
 
-            return nativeOverlapped;
+        private static unsafe IOCompletionCallback AllocateCallback()
+        {
+            return new IOCompletionCallback(Callback);
+
+            static unsafe void Callback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
+            {
+                ThreadPoolBoundHandle threadPoolBoundHandle = (ThreadPoolBoundHandle)ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped)!;
+                threadPoolBoundHandle.FreeNativeOverlapped(pOverlapped);
+            }
```

Comment:
Getting this working is the most important thing; thank you for taking the feedback about ensuring the sync and async operations always work regardless of the sync/async state of the handle.

Separately, we should do some perf analysis to determine whether this approach is the right one, vs just having the sync method delegate to the async method and block waiting for it to complete.  They each have their pros and cons.  The biggest benefit of this approach is we don't need to wait for the IO callback to be invoked by the thread pool thread in order for the sync operation to make forward progress; the callback is purely responsible for freeing the memory.  The downside is I think we might be allocating more per operation now, as each operation needs to allocate a ManualResetEvent, its SafeWaitHandle, and most impactfully the Overlapped and supporting structures; if it were delegating to the async operation, it might need to allocate a Task object, and potentially a ManualResetEventSlim, but I think those will be cheaper.  So, it'll take some analysis to get the right answer (unless you've already done that, in which case, thanks!).